### ISSUE TITLE
change in the mat64.Dot signature broke APIs in golearn/metrics

### ISSUE THREAD

**Issue by vineethrp** [sjwhitworth/golearn#146]
InnerProduct API is broken because there was an signature change for mat64.Dot() in gonum.

<pre>
vineeth@vinsys:~/go/src/github.com/sjwhitworth/golearn/examples/knnclassifier$ go build knnclassifier_iris.go 
# github.com/sjwhitworth/golearn/metrics/pairwise
../../metrics/pairwise/euclidean.go:17: cannot use vectorX (type *mat64.Dense) as type *mat64.Vector in argument to mat64.Dot
../../metrics/pairwise/euclidean.go:17: cannot use vectorY (type *mat64.Dense) as type *mat64.Vector in argument to mat64.Dot
../../metrics/pairwise/poly_kernel.go:21: cannot use vectorX (type *mat64.Dense) as type *mat64.Vector in argument to mat64.Dot
../../metrics/pairwise/poly_kernel.go:21: cannot use vectorY (type *mat64.Dense) as type *mat64.Vector in argument to mat64.Dot
</pre>

The fix would be something like
```
diff --git a/metrics/pairwise/euclidean.go b/metrics/pairwise/euclidean.go
index df865e3..999e2a2 100644
--- a/metrics/pairwise/euclidean.go
+++ b/metrics/pairwise/euclidean.go
@@ -14,7 +14,9 @@ func NewEuclidean() *Euclidean {
 
 // InnerProduct computes a Eucledian inner product.
 func (e *Euclidean) InnerProduct(vectorX *mat64.Dense, vectorY *mat64.Dense) float64 {
-       result := mat64.Dot(vectorX, vectorY)
+       subVector := mat64.NewDense(0, 0, nil)
+       subVector.MulElem(vectorX, vectorY)
+       result := mat64.Sum(subVector)
 
        return result
 }
diff --git a/metrics/pairwise/poly_kernel.go b/metrics/pairwise/poly_kernel.go
index 193351b..4f0c60b 100644
--- a/metrics/pairwise/poly_kernel.go
+++ b/metrics/pairwise/poly_kernel.go
@@ -18,7 +18,9 @@ func NewPolyKernel(degree int) *PolyKernel {
 // InnerProduct computes the inner product through a kernel trick
 // K(x, y) = (x^T y + 1)^d
 func (p *PolyKernel) InnerProduct(vectorX *mat64.Dense, vectorY *mat64.Dense) float64 {
-       result := mat64.Dot(vectorX, vectorY)
+       subVector := mat64.NewDense(0, 0, nil)
+       subVector.MulElem(vectorX, vectorY)
+       result := mat64.Sum(subVector)
        result = math.Pow(result+1, float64(p.degree))
 
        return result
```

**Comment by Sentimentron** [comment]
OK: so after a quick attempt at a fix, looks like some tests are failing. It might require a bit of a redesign of this particular API.

**Comment by Sentimentron** [comment]
This should now be resolved in PR #147. If it isn't, please update this issue.
